File: 05_dotNET/SrabRail_dotNET/0_projects_02/SrbRailFolderMonitor_17062009/CXmlReader.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.IO;

namespace SrbRailFolderMonitor
{
    // Klasa koja cita xml fajl, radi validaciju i poziva upis u bazu
    class CXmlReader
    {
        private CDB_SQL m_DB;

        public CDB_SQL DB
        {
            get { return m_DB; }
            set { m_DB = value; }
        }

        // Funkcija cita xml fajl i vrsi verifikaciju u skladu sa xsd fajlom.
        // poziva funkcije za upis u bazu (iz klase CDB_SQL)
        public int Load(XmlFileInfo fi, int idFolderSuccess, int idFolderFailed)
        {
            try
            {
                int id_file = 0;
                DateTime ftp_date;
                //int id_stat = 0;
                string country = "";
                string station = "";
                string operater = "";
                string consignment_note = "";
                long idParent = 0;
                XmlDocument doc = new XmlDocument();
                XmlReaderSettings settings = null;
                XmlDocument document = null;

                try
                {
                    CLog.Log(fi.sPath, "CXmlReader.Load.Validate");
                    settings = new XmlReaderSettings();
                    settings.Schemas.Add(null, fi.sSchemaFile);
                    settings.ValidationType = ValidationType.Schema;
                    document = new XmlDocument();
                    document.Load(fi.sPath);
                    XmlReader rdr = XmlReader.Create(new StringReader(document.InnerXml), settings);
                    while (rdr.Read())
                    {
                        ;
                    }
                    rdr.Close();
                }
                catch (Exception ex)
                {
                    CLog.Log(ex, "CXmlReader.LoadValidate");
                    m_DB.WriteXmlFile(ref id_file, Path.GetFileName(fi.sPath), fi.dtFtpDate, 4, fi.idFolder, idFolderFailed, "", "", "", "");
                    fi.idFile = id_file;
                    return 4;
                }


                CLog.Log("Write File: " + Path.GetFileName(fi.sPath) + " document:" + document.Name , "CXmlReader.Load.Write");
                XmlNodeList NL = document.GetElementsByTagName("versandbhf");
                foreach (XmlNode N in NL)
                {
                    //<vw> (zemlja)       <bhfnr> (stanica)       <evu> (prevoznik)      <versandnr> (broj tovarnog lista)
                    foreach (XmlNode n in N.ChildNodes)
                    {
                        if (n.Name == "vw") country = n.InnerText;
                        if (n.Name == "bhfnr") station = n.InnerText;
                        if (n.Name == "evu") operater = n.InnerText;
                        if (n.Name == "versandnr") consignment_note = n.InnerText;
                    }
               }
               /* WRITE TO DB*/
               /* Status je 3. Ako sve bude u redu sto je najcesce nema izmene sloga, samo se fajl pomeri sa direktorijuma */
               if (m_DB.WriteXmlFile(ref id_file, Path.GetFileName(fi.sPath), fi.dtFtpDate, 3, fi.idFolder, idFolderSuccess, country, station, operater, consignment_note) > 0)
               {
                   fi.idFile = id_file;

                   m_DB.BeginTransaction();
                   foreach (XmlNode n in document.ChildNodes)
                   {
                       if (!WriteNode(n, id_file, 1, idParent))
                       {
                           m_DB.Rollback();
                           return 4;
                       }
                   }
                   if (m_DB.Commit()) return 3;
                   else return 4;
               }
               else
               {
                   return 4;
               }
            }
            catch (Exception ex)
            {
                m_DB.Rollback();
                CLog.Log(ex, "CXmlReader.Load");
                return 4;
            }
        }

        // Rekurzivna funkcija koja treba da upisu slog u bazu i pozove funkciju za upis atributa ako ih nod poseduje
        private bool WriteNode(XmlNode N, long id_file, int deep, long idParent)
        {
            int pos = 0;
            string sNodeName = "";
            string sNodeValue = "";
            string sParentNodeName = "";
            long idNode = 0, idNodeType = 0;
            string sNodeType = "";
            bool bLast=false;

            try
            {
                deep++;
                if (deep > 100)
                {
                    CLog.Log("Deep > 100", "CXmlReader.WriteNode");
                    return false;
                }
                
                //string s = "Node: " + " deep=" + deep.ToString() + " Name=" + N.Name + " " + " Type=" + N.NodeType.ToString() + " ";
                sNodeName = N.Name;
                sParentNodeName = N.ParentNode.Name;

                /* WRITE TO DB*/
                //if (N.HasChildNodes)
                //CLog.Log("Node: " + N.Name + " Type=" + N.NodeType.ToString() + " deep=" + deep.ToString() + " ", "CXmlReader.WriteNode");
                pos = 1;
                sNodeType = N.NodeType.ToString();
                idNodeType = Convert.ToInt64(N.NodeType);
                switch (N.NodeType)
                {
                    case XmlNodeType.XmlDeclaration:
                        break;
                    case XmlNodeType.Text:
                        sNodeValue = N.InnerText;
                        bLast = true;
                        break;
                    case XmlNodeType.Element:
                        if (N.ChildNodes.Count == 1 && N.FirstChild.NodeType == XmlNodeType.Text)
                        {
                            sNodeValue = N.InnerText;
                            bLast = true;
                            break;
                        }
                        break;
                    case XmlNodeType.Comment:
                        break;
                    default:
                        break;
                }

                if (m_DB.WriteXmlNode(ref idNode, idParent, id_file, idNodeType, sNodeType, sNodeName, sNodeValue, sParentNodeName) > 0)
                {
                    pos = 2;
                    if (!bLast)
                    {
                        if (N.NodeType != XmlNodeType.XmlDeclaration && N.Attributes.Count > 0)
                        {
                            pos = 3;
                            if (!WriteAttributes(N, idNode)) return false;
                        }
                    }
                    pos = 4;
                    if (!bLast)
                    {
                        if (N.HasChildNodes)
                        {
                            pos = 5;
                            foreach (XmlNode n in N.ChildNodes)
                            {
                                pos = 6;
                                if (!WriteNode(n, id_file, deep, idNode)) return false;
                            }
                        }
                    }
                    return true;
                }
                else
                {
                    return false;
                }
            }
            catch (Exception ex)
            {
                CLog.Log(ex, "CXmlReader.WriteNode " + pos.ToString());
                return false;
            }
        }

        // Funkcija koja treba da upise atribute noda
        private bool WriteAttributes(XmlNode N, long idNode)
        {
            try
            {
                long idAttr = 0;
                foreach (XmlAttribute a in N.Attributes)
                {
                    /* WRITE TO DB*/
                    m_DB.WriteXmlAttribute(ref idAttr, idNode, a.Name, a.Value);
                    //CLog.Log("Attribute: " + a.Name + "=" + a.Value, "CXmlReader.WriteAttributes");
                }
                return true;
            }
            catch (Exception ex)
            {
                CLog.Log(ex, "CXmlReader.WriteAttributes");
                return false;
            }
        }
    }
}

File: 05_dotNET/SrabRail_dotNET/0_projects_02/FSMonitor/CXmlReader.vb
'''Imports System.Collections
'''Imports System.Text
'''Imports System.Xml
'''Imports System.IO


''''Public Class CXmlReader
'''Namespace SrbRailFolderMonitor
'''    ' Klasa koja cita xml fajl, radi validaciju i poziva upis u bazu
'''    Class CXmlReader
'''        Private m_DB As CDB_SQL

'''        Public Property DB() As CDB_SQL
'''            Get
'''                Return m_DB
'''            End Get
'''            Set(ByVal Value As CDB_SQL)
'''                m_DB = value
'''            End Set
'''        End Property

'''        ' Funkcija cita xml fajl i vrsi verifikaciju u skladu sa xsd fajlom.
'''        ' poziva funkcije za upis u bazu (iz klase CDB_SQL)
'''        Public Function Load(ByVal fi As XmlFileInfo, ByVal idFolderSuccess As Integer, ByVal idFolderFailed As Integer) As Integer
'''            Try
'''                Dim id_file As Integer = 0
'''                Dim ftp_date As DateTime
'''                'int id_stat = 0;
'''                Dim country As String = ""
'''                Dim station As String = ""
'''                Dim operater As String = ""
'''                Dim consignment_note As String = ""
'''                Dim idParent As Long = 0
'''                Dim doc As New XmlDocument
'''                Dim settings As XmlReaderSettings = Nothing
'''                Dim document As XmlDocument = Nothing

'''                Try
'''                    CLog.Log(fi.sPath, "CXmlReader.Load.Validate")
'''                    settings = New XmlReaderSettings
'''                    settings.Schemas.Add(Nothing, fi.sSchemaFile)
'''                    settings.ValidationType = ValidationType.Schema
'''                    document = New XmlDocument
'''                    document.Load(fi.sPath)
'''                    Dim rdr As XmlReader = XmlReader.Create(New StringReader(document.InnerXml), settings)
'''                    While rdr.Read()


'''                    End While
'''                    rdr.Close()
'''                Catch ex As Exception
'''                    CLog.Log(ex, "CXmlReader.LoadValidate")
'''                    m_DB.WriteXmlFile(id_file, Path.GetFileName(fi.sPath), fi.dtFtpDate, 4, fi.idFolder, idFolderFailed, _
'''                     "", "", "", "")
'''                    fi.idFile = id_file
'''                    Return 4
'''                End Try


'''                CLog.Log(("Write File: " & Path.GetFileName(fi.sPath) & " document:") + document.Name, "CXmlReader.Load.Write")
'''                Dim NL As XmlNodeList = document.GetElementsByTagName("versandbhf")
'''                For Each N__1 As XmlNode In NL
'''                    '<vw> (zemlja)       <bhfnr> (stanica)       <evu> (prevoznik)      <versandnr> (broj tovarnog lista)
'''                    For Each n__2 As XmlNode In N__1.ChildNodes
'''                        If n__2.Name = "vw" Then
'''                            country = n__2.InnerText
'''                        End If
'''                        If n__2.Name = "bhfnr" Then
'''                            station = n__2.InnerText
'''                        End If
'''                        If n__2.Name = "evu" Then
'''                            operater = n__2.InnerText
'''                        End If
'''                        If n__2.Name = "versandnr" Then
'''                            consignment_note = n__2.InnerText
'''                        End If
'''                    Next
'''                Next
'''                ' WRITE TO DB

'''                ' Status je 3. Ako sve bude u redu sto je najcesce nema izmene sloga, samo se fajl pomeri sa direktorijuma 

'''                If m_DB.WriteXmlFile(id_file, Path.GetFileName(fi.sPath), fi.dtFtpDate, 3, fi.idFolder, idFolderSuccess, _
'''                 country, station, operater, consignment_note) > 0 Then
'''                    fi.idFile = id_file

'''                    m_DB.BeginTransaction()
'''                    For Each n__2 As XmlNode In document.ChildNodes
'''                        If Not WriteNode(n__2, id_file, 1, idParent) Then
'''                            m_DB.Rollback()
'''                            Return 4
'''                        End If
'''                    Next
'''                    If m_DB.Commit() Then
'''                        Return 3
'''                    Else
'''                        Return 4
'''                    End If
'''                Else
'''                    Return 4
'''                End If
'''            Catch ex As Exception
'''                m_DB.Rollback()
'''                CLog.Log(ex, "CXmlReader.Load")
'''                Return 4
'''            End Try
'''        End Function

'''        ' Rekurzivna funkcija koja treba da upisu slog u bazu i pozove funkciju za upis atributa ako ih nod poseduje
'''        Private Function WriteNode(ByVal N__1 As XmlNode, ByVal id_file As Long, ByVal deep As Integer, ByVal idParent As Long) As Boolean
'''            Dim pos As Integer = 0
'''            Dim sNodeName As String = ""
'''            Dim sNodeValue As String = ""
'''            Dim sParentNodeName As String = ""
'''            Dim idNode As Long = 0, idNodeType As Long = 0
'''            Dim sNodeType As String = ""
'''            Dim bLast As Boolean = False

'''            Try
'''                deep += 1
'''                If deep > 100 Then
'''                    CLog.Log("Deep > 100", "CXmlReader.WriteNode")
'''                    Return False
'''                End If

'''                'string s = "Node: " + " deep=" + deep.ToString() + " Name=" + N.Name + " " + " Type=" + N.NodeType.ToString() + " ";
'''                sNodeName = N__1.Name
'''                sParentNodeName = N__1.ParentNode.Name

'''                ' WRITE TO DB

'''                'if (N.HasChildNodes)
'''                'CLog.Log("Node: " + N.Name + " Type=" + N.NodeType.ToString() + " deep=" + deep.ToString() + " ", "CXmlReader.WriteNode");
'''                pos = 1
'''                sNodeType = N__1.NodeType.ToString()
'''                idNodeType = Convert.ToInt64(N__1.NodeType)
'''                Select Case N__1.NodeType
'''                    Case XmlNodeType.XmlDeclaration
'''                        Exit Select
'''                    Case XmlNodeType.Text
'''                        sNodeValue = N__1.InnerText
'''                        bLast = True
'''                        Exit Select
'''                    Case XmlNodeType.Element
'''                        If N__1.ChildNodes.Count = 1 AndAlso N__1.FirstChild.NodeType = XmlNodeType.Text Then
'''                            sNodeValue = N__1.InnerText
'''                            bLast = True
'''                            Exit Select
'''                        End If
'''                        Exit Select
'''                    Case XmlNodeType.Comment
'''                        Exit Select
'''                    Case Else
'''                        Exit Select
'''                End Select

'''                If m_DB.WriteXmlNode(idNode, idParent, id_file, idNodeType, sNodeType, sNodeName, _
'''                 sNodeValue, sParentNodeName) > 0 Then
'''                    pos = 2
'''                    If Not bLast Then
'''                        If N__1.NodeType <> XmlNodeType.XmlDeclaration AndAlso N__1.Attributes.Count > 0 Then
'''                            pos = 3
'''                            If Not WriteAttributes(N__1, idNode) Then
'''                                Return False
'''                            End If
'''                        End If
'''                    End If
'''                    pos = 4
'''                    If Not bLast Then
'''                        If N__1.HasChildNodes Then
'''                            pos = 5
'''                            For Each n__2 As XmlNode In N__1.ChildNodes
'''                                pos = 6
'''                                If Not WriteNode(n__2, id_file, deep, idNode) Then
'''                                    Return False
'''                                End If
'''                            Next
'''                        End If
'''                    End If
'''                    Return True
'''                Else
'''                    Return False
'''                End If
'''            Catch ex As Exception
'''                CLog.Log(ex, "CXmlReader.WriteNode " & pos.ToString())
'''                Return False
'''            End Try
'''        End Function

'''        ' Funkcija koja treba da upise atribute noda
'''        Private Function WriteAttributes(ByVal N As XmlNode, ByVal idNode As Long) As Boolean
'''            Try
'''                Dim idAttr As Long = 0
'''                For Each a As XmlAttribute In N.Attributes
'''                    ' WRITE TO DB

'''                    'CLog.Log("Attribute: " + a.Name + "=" + a.Value, "CXmlReader.WriteAttributes");
'''                    m_DB.WriteXmlAttribute(idAttr, idNode, a.Name, a.Value)
'''                Next
'''                Return True
'''            Catch ex As Exception
'''                CLog.Log(ex, "CXmlReader.WriteAttributes")
'''                Return False
'''            End Try
'''        End Function
'''    End Class
'''End Namespace

''''End Class
